compiler/parser/parser.cpp

```diff
@@ -412,7 +412,7 @@ static void ensureRequiredStandardModulesAreParsed() {
       UnresolvedSymExpr* oldModNameExpr = toUnresolvedSymExpr(moduleExpr);
 
       if (oldModNameExpr == NULL) {
-        INT_FATAL("It seems an internal module is using a mod.submod form");
+        continue;
```

Comment:
This concerns me.  I mentioned on #15255 a concern about not parsing files when we need to as a result of skipping imports of nested modules, and this falls under the same category.  I think something is off with this function and would really like a closer look at how it is used and if there is a better way.  If we were in a re-entrant compiler world, I would not be as worried, as we could choose to bring in the files while resolving the `mod.submod` form

Reply:
This change allows a standard module to say `use this.SubModule`. Since this pattern cannot require additional parsing, it is not a big deal. Since this code only applies to standard modules, further problems with it in the field are unlikely, and once we have standard modules using other patterns (say submodules in separate files) we can update it according to what that pattern is.

Reply:
> Since this pattern cannot require additional parsing, it is not a big deal.

Yeah, but that's not the only pattern that will trigger the case.  A `toplevel.submodule` call will behave the same as a `this.submodule` or `super.submodule` call.  Unless you're putting in code to distinguish between those cases, you can't say that it doesn't require additional parsing.

> Since this code only applies to standard modules, further problems with it in the field are unlikely

Without this code, we're not going to know to come back to this function if further instances do come up (or at least it is going to be harder to remember we need to).

Reply:
@lydia-duncan: Would it be possible to write a test demonstrating the case that you're worried this change will break?  Overall, it feels better to me to enable code that we know to be correct (i.e., qualified uses/imports in standard modules) than to generate an internal error for such code given that the tests we added to ensure this routine works are still passing (particularly since I think of this code as being fairly corner-casey; whereas I think qualified uses/imports in standard modules aren't).  But maybe if I saw the case that you think the continue will cause to go off the rails I'd feel differently.

Reply:
Maybe, we'd have to use the "override location of standard/internal modules" flag to do it, though, since this is intentionally for internal modules

Reply:
Here's a test that demonstrates triggering the internal error today (the .good is what I would want to happen if we removed the `INT_FATAL`) https://github.com/chapel-lang/chapel/compare/master...lydia-duncan:testNestedInternal?expand=1

It's not a real case (most nested modules in the standard and internal directories already have their parent modules included by default, so won't run into this situation), but the module being referenced could reasonably be accessed on its own, I think.  Just more likely by user code than internal modules (due to it being a package module)

Reply:
Thanks for putting this together.  I see that we get the internal error on master today for this test whereas with compilers that replace the INT_FATAL with `continue;` we get:

```
useNestedDir/ChapelNumLocales.chpl:4: error: Cannot find symbol 'TOML'
```

But I'm not understanding why getting the internal error for this case would be preferable to not getting it in cases that _are_ reasonable (e.g., adding `import blah.blah;` to Sort.chpl).  It seems like we have an unimplemented feature in either case, but at present the internal error is preventing reasonable cases from working while getting the error above seems not particularly worse than the internal error (particularly since, as you note, it seems pretty unlikely to come up in practice), and maybe even slightly nicer (the compiler is saying more clearly where it's failing to do something).

So I'd be inclined to switch to the `continue`, and have you file your test as a future, creating an issue for it.  How does that sound?

Reply:
The internal error does have the benefit of indicating that this is a compiler bug rather than something the programmer should be required to fix, and leads you to the code that needs to be updated.

I'm okay with switching to the `continue`, but would like special casing so that we only continue if the first symbol is named "this" or "super" since we can match against those astrs pretty easily.  And maybe warn about other cases, indicating that a use or import of the top level may be required to find them?

Reply:
A warning would get us back to the right part of the code to modify much more quickly

Reply:
The challenge I see with adding a warning is that every program that's compiled would generate the warning now that Sort.chpl contains imports/uses that trigger this path.  For example, here's the output if a warning and list_view are added to the conditional for `hello.chpl`:

```
warning: Yo yo yo
257124  call( . Sort "defaultComparator" ) 
```

This conditional is also triggered for cases that don't rely on `super`/`this`—for example `import Sort.defaultComparator;` caused the warning above.

Reply:
I found that argument compelling.  Opened #15347